exporters/otlp/src/otlp_grpc_client.cc

```diff
@@ -46,16 +121,144 @@ static std::string GetFileContentsOrInMemoryContents(const std::string &file_pat
+    std::unique_ptr<google::protobuf::Arena> &&arena,
+    RequestType &&request,
+    std::function<bool(opentelemetry::sdk::common::ExportResult,
+                       std::unique_ptr<google::protobuf::Arena> &&,
+                       const RequestType &,
+                       ResponseType *)> &&result_callback,
+    int32_t export_data_count,
+    const char *export_data_name) noexcept
+{
+  if (async_data->running_requests.load(std::memory_order_acquire) >=
+      async_data->max_concurrent_requests)
+  {
+    OTEL_INTERNAL_LOG_ERROR("[OTLP GRPC Client] ERROR: Export "
+                            << export_data_count << " " << export_data_name
+                            << " failed, exporter queue is full");
+    if (result_callback)
+    {
+      result_callback(opentelemetry::sdk::common::ExportResult::kFailureFull, std::move(arena),
+                      request, nullptr);
+    }
+    return opentelemetry::sdk::common::ExportResult::kFailureFull;
+  }
+
+  std::shared_ptr<OtlpGrpcAsyncCallData<RequestType, ResponseType>> call_data =
+      std::make_shared<OtlpGrpcAsyncCallData<RequestType, ResponseType>>();
+  call_data->arena.swap(arena);
+  call_data->result_callback.swap(result_callback);
+
+  call_data->request =
+      google::protobuf::Arena::Create<RequestType>(call_data->arena.get(), std::move(request));
+  call_data->response = google::protobuf::Arena::Create<ResponseType>(call_data->arena.get());
+
+  if (call_data->request == nullptr || call_data->response == nullptr)
+  {
+    OTEL_INTERNAL_LOG_ERROR("[OTLP GRPC Client] ERROR: Export "
+                            << export_data_count << " " << export_data_name
+                            << " failed, create gRPC request/response failed");
+
+    if (call_data->result_callback)
+    {
+      call_data->result_callback(
+          opentelemetry::sdk::common::ExportResult::kFailure, std::move(call_data->arena),
+          nullptr == call_data->request ? request : *call_data->request, call_data->response);
+    }
+
+    return opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+  call_data->grpc_context.swap(context);
+
+  call_data->grpc_async_callback = [](OtlpGrpcAsyncCallDataBase *base_call_data) {
+    OtlpGrpcAsyncCallData<RequestType, ResponseType> *real_call_data =
+        static_cast<OtlpGrpcAsyncCallData<RequestType, ResponseType> *>(base_call_data);
+    if (real_call_data->result_callback)
+    {
+      return real_call_data->result_callback(real_call_data->export_result,
+                                             std::move(real_call_data->arena),
+                                             *real_call_data->request, real_call_data->response);
+    }
+
+    return true;
+  };
+
+  ++async_data->start_request_counter;
+  ++async_data->running_requests;
+#  if defined(GRPC_CPP_VERSION_MAJOR) && \
+      (GRPC_CPP_VERSION_MAJOR * 1000 + GRPC_CPP_VERSION_MINOR) >= 1039
+  stub->async()
+#  else
+  stub->experimental_async()
+#  endif
```

Comment:
What is the motivation for supporting GRPC < 1.39 ?

This seems very old, can we document that GRPC async requires a recent GRPC instead ?

If somehow this causes compile time breaks with OTLP HTTP async, how about to use a different compile time flag, like ENABLE_GRPC_ASYNC_EXPORT, to decouple HTTP async from GRPC async.

This will allow the OTLP HTTP implementation to become mainline sooner too, if we decide to do so.

Reply:
Ping - do we need to support GRPC < 1.39, or can we just document a recent GRPC is needed instead ?